com.unity.multiplayer.mlapi/Tests/Editor/ProfilerTests.cs

```diff
@@ -0,0 +1,192 @@
+using System;
+using System.Collections.Generic;
+using MLAPI.Profiling;
+using NUnit.Framework;
+using UnityEditor;
+using UnityEngine;
+
+namespace MLAPI.RuntimeTests
+{
+    public class TestTransport : ITransportProfilerData
+    {
+        internal static class ProfilerConstants
+        {
+            public const string TransportTestData = nameof(TransportTestData);
+        }
+
+        private static readonly ProfilingDataStore TransportProfilerData = new ProfilingDataStore();
+
+        public void BeginNewTick()
+        {
+            TransportProfilerData.Clear();
+        }
+
+        public IReadOnlyDictionary<string, int> GetTransportProfilerData()
+        {
+            return TransportProfilerData.GetReadonly();
+        }
+
+        public void Send(string testMessage)
+        {
+            PerformanceDataManager.Increment(ProfilerConstants.TransportTestData);
+        }
+    }
+
+    public class TestHasProfilable : IHasProfilableTransport
+    {
+        internal static class ProfilerConstants
+        {
+            public const string NetworkTestData = nameof(NetworkTestData);
+        }
+
+        private TestTransport m_Transport;
+
+        public ITransportProfilerData GetTransport()
+        {
+            return m_Transport;
+        }
+
+        public void Initialize(bool useNullTransport)
+        {
+            m_Transport = useNullTransport ? null : new TestTransport();
+            ProfilerNotifier.Initialize(this);
+        }
+
+        public static void ProfilerBeginTick()
+        {
+            ProfilerNotifier.ProfilerBeginTick();
+        }
+
+        public static void NotifyProfilerListeners()
+        {
+            ProfilerNotifier.NotifyProfilerListeners();
+        }
+
+        public void Send()
+        {
+            ProfilerNotifier.Increment(ProfilerConstants.NetworkTestData);
+            m_Transport?.Send("testMessage");
+        }
+    }
+
+    public class NoTickDataException : Exception
+    {
+    }
+
+    public class ProfilerTests
+    {
+        private static void BreakDownTestProfiler(bool useNullTransport)
+        {
+            if (useNullTransport)
+            {
+                ProfilerNotifier.OnPerformanceDataEvent -= TestProfilerOnPerformanceDataEventNoTransport;
+            }
+            else
+            {
+                ProfilerNotifier.OnPerformanceDataEvent -= TestProfilerOnPerformanceDataEventNormal;
+            }
+
+            ProfilerNotifier.OnNoTickDataEvent -= TestProfilerNotifierOnOnNoTickDataEvent;
+        }
+
+        private static TestHasProfilable SetupTestProfiler(bool useNullTransport)
+        {
+            ProfilerNotifier.OnNoTickDataEvent += TestProfilerNotifierOnOnNoTickDataEvent;
+            if (useNullTransport)
```

Comment:
I see that we're wiring up a [custom assertion](http://xunitpatterns.com/Custom%20Assertion.html) here in the test profiler setup, that is called later on. 

An important quality for a good tests is obviously readability. For example, I want my test to be clear about the [Given/When/Then pattern](https://martinfowler.com/bliki/GivenWhenThen.html) so that it's easy to understand where these things happen. For the "Then" part of that pattern, we expect to see an assert to be called so it's clear how we're failing the test.

For this reason, I wonder if we should consider not wiring up the TestProfilerOnPerformanceDataEventNormal and TestProfilerOnPerformanceDataEventNoTransport custom assertion method this way, and instead include them as explicit calls in the test? It's not as cool of a programming technique as adding them to an event handler for sure, but I think it elevates the readability of the test by using this approach:

// Given
...
// When
...
// Then
AssertNetworkAndTransportDataExist // calling it here in the test makes it clear how we consider a failure in the test.